clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -416,7 +416,8 @@ class DiagDeviceFunction : public RecursiveASTVisitor<DiagDeviceFunction> {
       // instantiation as template functions. It means that
       // all functions used by kernel have already been parsed and have
       // definitions.
-      if (RecursiveFuncs.count(Callee) && !ConstexprDepth) {
+      if (!Callee->isConstexpr() && RecursiveFuncs.count(Callee) &&
```

Comment:
I don't think this is right.  This is only checking that the callee is a constexpr function, but this doesn't say that we're requiring that the call be constexpr.  Consider:

````
constexpr int foo(){ return 0; }

void bar() {
  constexpr int I = foo(); // we don't want to diagnose THIS.
  SomeTemplate<foo()>(); // we also don't want to diagnose this.
  int J = foo(); // We DO want to diagnose this, since it isn't being called in a constant evaluated context.


````

Reply:
> I don't think this is right. This is only checking that the callee is a constexpr function, but this doesn't say that we're requiring that the call be constexpr. Consider:
> 
> ```
> constexpr int foo(){ return 0; }
> 
> void bar() {
>   constexpr int I = foo(); // we don't want to diagnose THIS.
>   SomeTemplate<foo()>(); // we also don't want to diagnose this.
>   int J = foo(); // We DO want to diagnose this, since it isn't being called in a constant evaluated context.
> ```

I don't understand what's going on here. I've added a bunch of printf to see what's happening. I see the called function being added to RecursiveFunctions. I added a loop to dump out the RecursiveFunctions in the destructor for DeviceFunctionTracker. The called function is in RecursiveFunctions and the "count" is 1. However in VisitCallExpr the "count" is 0. How does it go to 0? I don't see erase getting called. Any clue?   In this expression "ConstexprDepth" is non-zero when it is in a constexpr usage context, is that right?

Reply:
I wouldn't expect anything to be removed from the RecursiveFunctions list, that is correct.   

Presumably it got copied at one point, because we entered the 'if' statement to diagnose, right?

The constexpr depth _IS_ supposed to be incremented when entering a constexpr context, but I don't know why it wouldn't.  I would expect the 'TraverseVarDecl' to have worked for it in that case.

Reply:
For now i'm looking at "count" and it's zero even tho' it is or was in the RecursiveFuncs

Reply:
Could it be some sort of canonicalization problem?  I thought we were careful about making sure we always inserted the canonical decl, but perhaps we missed one?  Or it missed it here?

Reply:
The FD node in the set is 0x140acf30 but the FD node in the CallExpr is 0x140acd98 so that's why the count is 0. The code is using getCanonicalDecl() both to enter it in the RecursiveFun and also to check if it's in RecursiveFun, is something else needed?

Reply:
OK I'll double back and check canonical everywhere

Reply:
I THINK canonicalization everywhere should work, and both checks should work fine, just not sure why it wouldn't work.  CanonicalDecl does 'getFirstDecl' for FunctionDecl, so it should always be the first one...

Reply:
o I'm not sure what you're seeing with the FD node, but I also am not sure why it is important here... 

But the issue is that the Diagnoser is visiting the function-decls 1 at a time (we always did, so I don't know why this is a regression, other than we are now doing a 'better' job detecting recursion).

So basically we call CheckBody on every function, and by the time you get to the recursive function,it doesn't know it was used constexpr-ly.

```
  constexpr int foo(int I) {
    if (I != 0)
      return foo(I - 1);
    return I;
  }
 
  template <typename T, typename Func>
  __attribute__((sycl_kernel)) void kernel(const Func &f) {
    f();
  }
 
  int main() {
    kernel<class F>([]() {
      constexpr int I = foo(5);
      return I;
    });
  }
```

So we visit the lambda `operator()` first, but aren't doing the 'call descend' in that same context.  We later visit the `foo` function, where it recognizes it is called recursively, but not that the initial call to it was in a `constexpr` context.

Unless someone has a better idea, I only have 1 or 2 solutions:
1- Add an option to the `CallGraph`/`RecursiveASTVisitor` type in some way to make it no longer 'descend' into constant expressions.  This will probably suppress some of the other diagnostics we have, IMO, correctly (like `VisitCXXTypeidExpr` or `VisitCXXDynamicCastExpr`, or  virtual functions, or `DLLImport`, or the math library checks, for the function pointer checks).

2- Change the `CallGraph` to have its `CallGraphNode` (or perhaps the `CallGraphNode::CallRecord` type?) store whether it is in a constant evaluated context.  I'm not sure this one is viable, since functions might be called both ways, and I don't know the implications of that.

Reply:
Another, perhaps 'better', option I thought of a moment ago...  Inherit from `CallGraph`, have the `DeviceFunctionTracker` use THAT type instead.  In THAT type, override the Traverse** things to just `return true` (suppressing going deeper into them) in the same cases we do the `ConstexprDepthRAII` in this current type.

Reply:
The reason the FD node is important is because we cannot detect recursion. Count is 0 for the recursive function.

Reply:
I thought the problem you're trying to solve is diagnosing a recursive function when we SHOULDN'T be (thanks to constant evaluation contexts?)?

Reply:
^^Also note, I did a 'dumb' in my 1st example, note that 'foo' is not actually recursive...